### PR TITLE
Render neighboring inline tags with no whitespace between them

### PR DESCRIPTION
**Pull request by wch**
This PR makes the following changes.

Each tag function gets a boolean parameter `_add_ws`, which defaults to `False` for (normally) inline components like `<span>`, and `True` for (normally) block components like `<div>`. The user _can_ call those functions with a different value, because it is possible for CSS to change a component from inline to block and vice versa.

If two inline tags (like `<span>` or plain strings), are next to each other, they will be rendered with no whitespace between them:

```html
span(span("a"), span("b"))
<span><span>a</span><span>b</span></span>

span("a", "b")
<span>ab</span>
```

The examples above show what happens if the inline tags are in another inline tag. Note that there's no whitespace between the opening `<span>` tag and the first child.

If they are within a block tag like `div`, there will be whitespace between the opening `<div>` tag and the first child.

```
div(span("a"), span("b"))
<div>
  <span>a</span><span>b</span>
</div>


div("a", "b")
<div>
  ab
</div>
```

If a block tag contains a block tag, each of the block children will be on its own line. If there are multiple inline children in sequence, they will render without whitespace between them.

```
div(div("a"), span("b"), span("c"), div("d"))
<div>
  <div>a</div>
  <span>b</span><span>c</span>
  <div>d</div>
</div>
```

I have not described the behavior when an inline tag contains a block tag, like `span(div("a"))`. This is because in normal HTML, block tags can contain inline or block tags, but inline tags can only contain inline tags -- inline tags cannot contain block tags.

I added many tests for the correct whitespace behavior.

Note: One important change is that when a user writes code like this, no whitespace will be added automatically:

```py
div("Hello", "world!")
```

Previously, that would render as:

```html
<div>
  Hello world!
</div>
```

Now it will render as:

```html
<div>
  Helloworld!
</div>
```

This is a significant change, but it's best to do it now rather than later.

### CODE REVIEW

File: htmltools/_core.py [hunk 1]
```diff
@@ -242,7 +242,12 @@ def render(self) -> RenderedHTML:
         return {"dependencies": deps, "html": cp.get_html_string()}
 
     def get_html_string(
-        self, indent: int = 0, eol: str = "\n", *, _escape_strings: bool = True
+        self,
+        indent: int = 0,
+        eol: str = "\n",
+        *,
+        add_ws: bool = True,
```
Comment (cpsievert) [inline htmltools/_core.py:249]: Add to docstring

File: scripts/generate_tags.py [hunk 2]
```diff
@@ -7,8 +7,51 @@
 import re
 from urllib.request import urlopen
 
+_INLINE_TAG_NAMES = {
```
Comment (cpsievert) [inline scripts/generate_tags.py:10]: Looks like you got this list from #34, but I don't remember where I got that list (sorry), and this seems that there are more tags to consider adding here https://developer.mozilla.org/en-US/docs/Web/HTML/Inline_elements#list_of_inline_elements

File: htmltools/_core.py [hunk 3]
```diff
@@ -434,6 +465,26 @@ class Tag:
     children
         The tag's children.
 
+    Note
+    ----
+    The `_add_ws` parameter controls whether whitespace is added around the tag. By
+    default, tag functions for block elements (like `div()` and `p()`) set this to
+    `True`, and tag functions for inline elements (like `span()` and `a()`) set this to
+    `False`.
```
Comment (cpsievert) [inline htmltools/_core.py:473]:
```suggestion
    The `_add_ws` parameter controls whether whitespace is added around the tag. 
    Inline tags (like `span()` and `a()`) default to  `False` and block tags 
    (like `div()` and `p()`) default to `True`.
```

File: htmltools/_core.py [hunk 4]
```diff
@@ -434,6 +465,26 @@ class Tag:
...
+
+    When a tag with `_add_ws=True` is rendered to HTML, whitespace (including
+    indentation) is add before the opening tag (like `<div>`), after the closing tag
```
Comment (cpsievert) [inline htmltools/_core.py:476]:
```suggestion
    indentation) is added before the opening tag (like `<div>`), after the closing tag
```

File: tests/test_tags.py [hunk 5]
```diff
@@ -179,29 +175,226 @@ def test_tagify_deep_copy():
...
+            """\
+            <div>
+              <div>
+                <span>a</span>
+              </div>
+              <span><b>b</b></span>
+            </div>"""
+        ),
+    )
+
```
Comment (cpsievert) [inline tests/test_tags.py:371]: Maybe it'd be worth adding some test cases for `TagList()` (e.g., `TagList("a", "b", div("c", "d"), span("e", "f"), span("g", "h"))`)?

**Review by jcheng5 — APPROVED** [review]

**Pull request merged**